Implement bar texts (issue 34)

n-riesco opened the pull request:
This PR implements attributes `textposition`, `textfont`, `insidetextfont` and `outsidetextfont` for bar traces.

I've followed [this comment](https://github.com/plotly/plotly.js/issues/34#issuecomment-160645382) as close possible:
- all these attributes accept arrays
- default `textposition` is `none`
- if `textposition` was set to `auto`, the `inside` options is preferred over `outside`
- if `outside` was requested, but the bar is stacked under another bar, then the text is drawn `inside`
- if `inside` was requested, but the bar is too small, then no text is drawn

I've also added jasmine tests and mocks exercising the new functionality.

n-riesco committed:
bar: add text-related attributes

* Added attributes textposition, textfont, insidetextfont and
  outsidetextfont.

* All these attributes accept arrays.

n-riesco committed:
bar: draw bar texts

n-riesco committed:
test: bar texts

n-riesco committed:
test: add mocks using bar texts

alexcjohnson commented on src/traces/bar/plot.js:
```diff
@@ -42,9 +45,9 @@ module.exports = function plot(gd, plotinfo, cdbar) {
+            d3.select(this).selectAll('g.point')
                 .data(Lib.identity)
-              .enter().append('path')
+                .enter().append('g').classed('point', true)
```
The funny indentation this had is fairly standard with d3, to mark operations within a chain that change the selection. Not sure how consistent we've been with that but it was done that way intentionally.

alexcjohnson commented on src/traces/bar/defaults.js:
```diff
@@ -33,7 +155,23 @@ module.exports = function supplyDefaults(traceIn, traceOut, defaultColor, layout
+        if(hasOutside) {
+            coerceArray('outsidetextfont', coerceFont, textFont, layout.font);
+        }
+    }
```
It looks like you've based this on [pie/defaults](https://github.com/plotly/plotly.js/blob/master/src/traces/pie/defaults.js#L51-L62) but you've got more logic around arrays in the input. So two questions:
- for the cases that _are_ covered by pie, does this give the same inheritance from `layout.font` and `textfont` to `insidetextfont` and `outsidetextfont`, particularly around cases where parts of the font come from one default and other parts come from the other one? for example if `layout.font` supplies the family, `textfont` supplies the size, and `insidetextfont` makes it white.
- what extra cases does this cover and should we extend pie to cover these too?

n-riesco replied:
Yes, I've tried to replicate the behaviour of pie traces. This is tested [here](https://github.com/n-riesco/plotly.js/blob/2526e83c58cc2203d9336309d8de7d1ab082a9de/test/jasmine/tests/bar_test.js#L120-L177). The test is a good example of the extra cases handled here.

The main difference is that these attributes now also accept arrays (so that fonts can be defined per bar).

I think a lot of the functionality implemented in this file could be moved to `Lib` and reused by pie traces.

alexcjohnson replied:
Cool. Looks like the inheritance for non-arrays is as I'd expect 👍 . For arrays though, I'm not sure about the structure - I'd think for consistency with data attributes and src tags, it would be better to have family, size, and color be arrays, rather than have font be an array of objects.

etpinard replied:
We don't coerce entries inside data arrays or `arrayOk` attributes.

Non-text entries are usually via fallbacks later on in the pipeline e.g. in the drawing step for scatter `text` [here](https://github.com/plotly/plotly.js/blob/d3c59da02d4451842d56e5e933e0e13090af9171/src/components/drawing/index.js#L350).

The reason being: we try to make `Plots.supplyDefaults` scale ~ number of attribute as opposed to ~ length of the data.

etpinard replied:
What the pie trace type is currently doing ([here](https://github.com/plotly/plotly.js/blob/d3c59da02d4451842d56e5e933e0e13090af9171/src/traces/pie/defaults.js#L51-L62)) looks sufficient to me at the default step.

n-riesco replied:
@etpinard I don't understand the phrase `"to make Plots.supplyDefaults scale ~ number of attribute"`? Could you give me an example?

etpinard replied:
As in: there are about as many coerce statements as there are attributes.

What you added above has about as many coerce statements as there are items in the `text` attribute arrays.

n-riesco replied:
@etpinard What are the reasons against making `Plots.supplyDefaults` scale with the data? Note that the calls to `nestedProperty` would still scale only with the number of attributes.

Having uncoerced data in arrays after running `Plots.supplyDefaults` means that we have to coerce this data in `Bar.plot`. This increases the number of places where untrusted user input can brake Plotly.

Another disadvantage of moving the code for array coercion from `Bar.supplyDefaults` into `Bar.plot` (instead of `Lib`)  is that pie traces wouldn't be able to reuse the code.

etpinard replied:
Perf over DRY.

`supplyDefaults` gets called a lot more often than `.plot`.

n-riesco replied:
@etpinard OK, I understand. Thx for the explanation.

alexcjohnson commented on src/traces/bar/plot.js:
```diff
@@ -108,12 +115,274 @@ module.exports = function plot(gd, plotinfo, cdbar) {
+
+    if(textPosition === 'inside') {
+        if(barIsTooSmall) return;
+    }
```
For pies we did this a bit differently - in `'inside'` mode we would always draw the text inside the slice, and shrink it if need be so it would fit (no matter how small the space is). And in `'auto'` mode we would first draw it inside, then if it had to be scaled down to fit inside we'd move it outside.

This algorithm should be easier with a rectangular bar than with a wedge-shaped pie slice - do you want to give it a try?

n-riesco replied:
Yes, no problem. One question though. If a text doesn't fit, shall I consider rotating the text before moving it outside?

alexcjohnson replied:
> If a text doesn't fit, shall I consider rotating the text before moving it outside?

Yes, I think so. Basically do everything you'd do with explicitly inside text, which includes seeing if you can make it bigger by rotating it, and only after that see if you can make it bigger by moving it outside. I guess even outside text can get squished here if it's wider than the bar... so it would be silly to move the text outside and still not be able to expand it!

alexcjohnson replied:
makes me wonder if we need an attribute `textorientation` or something, that would be `'auto'` (whatever is biggest), `'horizontal'`, or `'vertical'`.

etpinard replied:
> I consider rotating the text before moving it outside?

I think so, as long as the text can only be horizontal or vertical (no diagonal allowed).

n-riesco replied:
@alexcjohnson I've set a 3px padding. This means that unless a bar is larger than 6px, no labels could be drawn inside. Is it OK not  to draw the label in those cases? Or shall I remove the padding if the bar is 6px or less?

etpinard replied:
FYI @alexcjohnson will be offline until Monday.

etpinard replied:
6px sounds ok to me.

alexcjohnson replied:
An alternative that might be more analogous to what we did with pies (where we didn't explicitly include padding, but fitting text into a rectangle into a wedge effectively results in some padding in most cases) would be to scale down the padding by the same factor you use to scale down the label. The only context I see this making a big difference is print / pdf, where hover & zoom are unavailable but the interested viewer may be able to zoom in the document (or hold it up to their nose) and still read it.

Not blocking - we can always tweak this later if it becomes important - but I think it would be a bit nicer that way.

alexcjohnson commented on test/image/mocks/bar_text_relative.json:
```diff
@@ -0,0 +1,40 @@
+            "x":[1,2,3,4],
+            "type":"bar"
+        }, {
+            "text":[0,"very small size",-3,-2],
```
is there a criterion by which we should rotate an outside label ('very small size') to vertical like you do with inside labels ('Three')?

Also, and this applies to all the inside labels, I'd have expected inside labels to be pushed next to the bar end, the same way as you have outside labels right on the other side of the bar end. It looks like [some people](http://stackoverflow.com/questions/6644997/showing-data-values-on-stacked-bar-chart-in-ggplot2) want the labels in the middle, so perhaps this should be an option (perhaps textalign: middle or end) but I think semantically it's a better default to draw people's eyes toward the bar end. Seems like that's [somewhat more common](https://www.google.com/search?q=bar+chart+with+text&espv=2&biw=1568&bih=842&source=lnms&tbm=isch&sa=X&ved=0ahUKEwjDnuKj6pnQAhUO3GMKHf8HCIIQ_AUIBigB&dpr=1#tbm=isch&q=bar+chart+with+text+inside) as well.

n-riesco replied:
@alexcjohnson Without going into the details, the idea is that if a text doesn't fit inside a bar, we shrink the text by computing a scaling factor (smaller than 1).

To avoid shrinking the text unnecessarily or too much, the text is rotated 90 degrees when this results in a larger scaling factor (still smaller than 1).

---

I'm not sure if I did a good job with the above explanation. If not, I can give an example.

alexcjohnson replied:
Yes, that's the correct behavior for inside labels, and it's basically the same thing we do for pies.

But there were two additional issues in my comment:
- for _outside_ labels, shall we do the same thing? ie rotate the bar 90 degrees when it would allow us to enlarge the scaling?
- for _inside_ labels, once we've figured out the maximum scaling, can we push the label over near the end of the bar rather than the middle?

n-riesco replied:
@alexcjohnson I've just realised that my implementation for `outside` labels doesn't do what I said :disappointed: . How do you think it should be? Shall I rotate it like in the case of `inside` labels?

---

:+1: Yes, I'll push the labels close to the bar ends. One question: at the moment I'm using a 1px padding, shall I increase it (the padding in [the ggplot2 example](http://stackoverflow.com/questions/6644997/showing-data-values-on-stacked-bar-chart-in-ggplot2) looks too much, though)?

alexcjohnson replied:
> Shall I rotate it like in the case of inside labels?

Yes, I think so. One concern with this (it's already a concern but will get worse when we rotate the text) is whether we need to include that label in autorange (via `Axes.expand` and its `ppadplus` and `ppadminus` options). That could get awfully annoying though, as you'd have to make the outside text elements available for sizing during `setPositions` rather than `plot`, but then you still need to know the bar width to tell whether to rotate or not... ugh. @etpinard thoughts?

> 1px padding, shall I increase it (the padding in the ggplot2 example looks too much, though)?

Agreed about ggplot2. For hover text [we use 3px](https://github.com/plotly/plotly.js/blob/master/src/plots/cartesian/constants.js#L64) - how about that?

etpinard commented on src/plot_api/helpers.js:
```diff
@@ -299,7 +299,7 @@ exports.cleanData = function(data, existingData) {
         }
 
-        if(!Registry.traceIs(trace, 'pie')) {
+        if(!Registry.traceIs(trace, 'pie') && !Registry.traceIs(trace, 'bar')) {
```
no need to clean (plotly.js lingo for backward-compatible mappings) attribute that have never been released.

n-riesco replied:
That's what this line change is meant for; i.e. to exclude bar's `textposition` from those mappings.

etpinard replied:
Oops. You're right. My mistake. :smile:

etpinard commented on src/traces/bar/set_positions.js:
```diff
@@ -179,6 +179,24 @@ function setGroupPositionsInStackOrRelativeMode(gd, pa, sa, calcTraces) {
+            if(!isNumeric(bar.s)) continue;
+
+            var isOutmostBar = ((bar.b + bar.s) === sieve.get(bar.p, bar.s));
+            if(isOutmostBar) bar._outmost = true;
```
nicely done here.

etpinard commented on src/traces/bar/set_positions.js:
```diff
@@ -179,6 +179,24 @@ function setGroupPositionsInStackOrRelativeMode(gd, pa, sa, calcTraces) {
     stackBars(gd, sa, sieve);
+
+    // flag the outmost bar (for text display purposes)
+    for(var i = 0; i < calcTraces.length; i++) {
```
We don't need to run this loop if `textposition: 'none'`, correct?

n-riesco replied:
Last night, you made me hesitate! But after further thought, here's the correct reasoning.

Bars with `textposition: 'none'` are still visible (only the bar text is hidden) and thus they can be the outmost bar.

etpinard commented on test/image/mocks/bar_text_horizontal_group.json:
```diff
@@ -0,0 +1,23 @@
+{
+    "data":[
```
@n-riesco general comment: would mind adding the `text` attributes and friends to the existing mocks you added for `width` / `offset` / `base`? We have over 300 image mocks now - which take about 2 and 1/2 minute to tests. It's time to be more meticulous when adding new mocks.

n-riesco replied:
I should've thought of that.

etpinard commented on test/jasmine/tests/bar_test.js:
```diff
@@ -612,6 +702,168 @@ describe('A bar plot', function() {
+        expect(pathBB.left).not.toBeGreaterThan(textBB.left);
+        expect(textBB.right).not.toBeGreaterThan(pathBB.right);
+        expect(pathBB.top).not.toBeGreaterThan(textBB.top);
+        expect(textBB.bottom).not.toBeGreaterThan(pathBB.bottom);
```
very nice here.

Asserting `getBoundingClientRect` return values would be very susceptible to cross-browser cross-env discrepancies.

n-riesco replied:
Since I've used `Drawing.bBox(node)` to compute the location of bar texts and `Drawing.bBox(node)` uses `getBoundingClientRect` this shouldn't be an issue.

I noticed, however, in certain cases, a rounding error that would cause the bar text to be negligibly larger than computed. For that reason, and also because it looks neater, I added some padding to bar texts. See [here](https://github.com/n-riesco/plotly.js/blob/7d7017c5b97b7204a3e74ef44d3d03de720d2be7/src/traces/bar/plot.js#L279-L281) and [here](https://github.com/n-riesco/plotly.js/blob/7d7017c5b97b7204a3e74ef44d3d03de720d2be7/src/traces/bar/plot.js#L298-L323).

etpinard commented:
> all these attributes accept arrays

very nice

> default textposition is none

yes :+1: 

> if textposition was set to auto, the inside options is preferred over outside

I vote :+1: on that

> if outside was requested, but the bar is stacked under another bar, then the text is drawn inside
> if inside was requested, but the bar is too small, then no text is drawn

answered in https://github.com/n-riesco/plotly.js/pull/4#pullrequestreview-7683700

@n-riesco I'm pretty excited about this. Once again, you'll make a bunch of users happy!

n-riesco committed:
bar: fix indentation of .enter()

n-riesco committed:
bar: increase text padding to 3px

n-riesco committed:
test: bar text nodes exist

n-riesco committed:
bar: change criteria for `textposition: 'auto'`

* With the new criteria, bar texts are drawn inside bars except in those
  cases that would require scaling down.

n-riesco committed:
test: update mock bar_text_relative

* Added rotated and shrunk inside text.

* Shortened outside texts.

n-riesco committed:
bar: update algorithm to position bar texts

* Inside texts are rotated only if the rotation helps shrink the text as
  little as possible.

* Outside texts are rotated only if the rotation doesn't leave the text
  perpendicular to the bar.

n-riesco committed:
bar: don't apply text padding if bar too small

* This change ensures inside texts can be drawn, even inside bars
  smaller than the text padding.

n-riesco committed:
test: remove mock bar_text_relative

* Updated mock bar_attrs_relative to test the bar test functionality.

n-riesco committed:
test: remove mock bar_text_horizontal_group

* Updated mock bar_attrs_group_norm to test the bar text functionality.

n-riesco committed:
test: remove bar_text_horizontal_relative_percent

* Updated the jasmine tests to test whether outside bars are correctly
  identified when `barnorm: 'percent'`.

n-riesco committed:
test: remove mock bar_text_horizontal_relative

n-riesco committed:
test: update mock bar_attrs_relative

* Reduced font size of text "Three" to ensure the baseline image also
  fits the text inside the bar.

n-riesco committed:
test: update mock bar_attrs_group_norm

* Set outside font size to 32 to force rotation. This change will test
  that horizontal bars rotate the bar text clockwise or anti-clockwise
  as appropriate.

n-riesco committed:
bar: fix normalised group bar plots

* Fixed bug computing the normalisation of bar plots in group mode.

* The bug was caused by calling `sieve.put` with bar sizes that didn't
  take into account the bar base.

* Bar plots in stack and relative mode weren't affected by this bug,
  because traces that set a bar base are excluded from the
  stack/relative mode.

n-riesco committed:
test: update mock bar_attrs_group_norm

* Updated the mock bar_attrs_group_norm so that it exercises the
  functionality to rotate bar texts.

n-riesco committed:
bar: fix hover label

* Fix hover xLabelVal and yLabelVal so that it takes into account the
  bar base.

n-riesco committed:
test: update hover tests in bar_test

* Updated hover tests in bar_test to test that hover labels honor bar
  bases.

n-riesco committed:
bar: update text rotation algorithm

* If rotation is needed:
  - inside texts are rotated counter-clockwise.
  - outside texts are rotated clockwise or counter-clockwise, so that
    the text is laid along the bar end.

n-riesco committed:
test: update baseline images

* Removed the baselines for bar_test_* mocks.

* Updated bar_attrs_relative and bar_attrs_group_norm that now exercise
  the new functionality to draw bar texts.

n-riesco committed:
test: make linter happy

n-riesco committed:
bar: do not coerce elements of an array attribute

n-riesco committed:
bar: accept font family, size and color arrays

* Before this commit, font attributes accepted:
  [{family: family0, size: size0, color: color0},
   {family: family1, size: size1, color: color1},
   ...]

* With this commit, font attributes accept:
  {family: [family0, family1, ...],
   size: [size0, size1, ...],
   color: [color0, color1, ...]}

n-riesco committed:
test: update bar font coercion tests

n-riesco committed:
test: update mock bar_attrs_relative

* Updated mock bar_attrs_relative to account for the change in the
  way to declare fonts.

n-riesco commented:
@etpinard @alexcjohnson 

The PR is ready for review again:
- I've moved the code to coerce array elements from `Bar.supplyDefaults` to `Bar.plot`.
- I've changed the attributes `textfont`, `insidetextfont` and `outsidetextfont` so that now they accept arrays in `family`, `size` and `color`.
- I've revamped the algorithm to position bar labels:
  - inside labels:
    - are drawn near the corresponding bar end
    - with a 3px padding (if the bar is too small, no padding is used)
    - are rotated counter-clockwise if this results in a larger label
  - outside labels:
    - are rotated clockwise or counter-clockwise so that they are laid along the bar end
    - laying the label along the bar end can result in tiny labels
    - I think we should rotate outside labels, in those case where otherwise they would be shrunk beyond a certain limit

The following mocks show best how the new positioning algorithms work:

![](https://raw.githubusercontent.com/n-riesco/plotly.js/a9a13f156737f5792d681e187582fb70a6040c09/test/image/baselines/bar_attrs_relative.png)

![](https://raw.githubusercontent.com/n-riesco/plotly.js/a9a13f156737f5792d681e187582fb70a6040c09/test/image/baselines/bar_attrs_group_norm.png)

etpinard commented:
@n-riesco great. I'll give this a look in the next few hours. Thanks!

etpinard commented on src/traces/bar/plot.js:
```diff
@@ -108,12 +125,430 @@ module.exports = function plot(gd, plotinfo, cdbar) {
+    return transformTranslate + transformScale + transformRotate;
+}
+
+function getText(trace, index) {
```
Great. Nice and clear.

This is pretty much on bar with the [`Drawing`](https://github.com/plotly/plotly.js/blob/master/src/components/drawing/index.js) does for scatter traces. So :+1: 

That said, I believe (I haven't tried yet) that some of this logic could be moved to the `calc` (which is called even less often than `plot` and `supplyDefaults`). But, that will be for a different PR.

n-riesco replied:
@etpinard Is there anything in `Drawing` that you think I should've reused?

etpinard replied:
`Drawing` is pretty messy at the moment. Though it would be nice to merge the text-drawing logic from scatter, pie and bar together at some point. No action required for this iteration.

etpinard commented on src/traces/bar/plot.js:
```diff
@@ -108,12 +125,430 @@ module.exports = function plot(gd, plotinfo, cdbar) {
+    return value;
+}
+
+function coerceString(attributeDefinition, value, defaultValue) {
```
Can you simply reuse [`Lib.coerce`](https://github.com/plotly/plotly.js/blob/master/src/lib/coerce.js) here?

n-riesco replied:
I couldn't do it when `textfont` accepted an array, but now I should be able to.

etpinard replied:
:arrow_up: **blocking** before making a PR to plotly.js

n-riesco replied:
@etpinard After some sleep, I remembered that the reason why I introduced those functions is that I don't know how to use `Lib.coerce` to coerce array elements. Even if something like `Lib.coerce('textfont.family[' + index + ']', dflt)` works, iterating an array like this must be awfully inefficient.

Would it be OK for me to modify [src/lib/corce.js](https://github.com/plotly/plotly.js/blob/e94b1d16f99514f09309e8ac2c294fdfdd1ddefb/src/lib/coerce.js#L92-L111) and move `coerceString` to a `coerce` field and update the corresponding [coerceFunction](https://github.com/plotly/plotly.js/blob/e94b1d16f99514f09309e8ac2c294fdfdd1ddefb/src/lib/coerce.js#L101) to use this new coerce field?

etpinard replied:
> iterating an array like this must be awfully inefficient.

Very true. That's why we don't use `Lib.coerce` on array items usually. You probably noticed that `Drawing` doesn't have any `coerce` statements.

For what it's worth, here's an example of how to get it to work with array under [`info_array`](https://github.com/plotly/plotly.js/blob/e94b1d16f99514f09309e8ac2c294fdfdd1ddefb/src/lib/coerce.js#L237-L251).

So, I'll leave it up to you. I'm ok with the current solution. But ideally, we should move all array-item casting to the `calc` step - which can wait for a refactor PR.

etpinard commented on test/image/mocks/bar_attrs_group_norm.json:
```diff
@@ -1,19 +1,24 @@
-            "x":[3,2,1,0],
+            "text":0.5,
+            "textposition":"outside",
+            "outsidetextfont": {"size":32},
```
Hmm. Looks like that `"size": 32` isn't reflected on the baseline image.

n-riesco replied:
@etpinard This is on purpose to force a rotation in all the outside labels. If the font is small enough, the label can be drawn without rotation.

You can play with the size to see what I mean.

etpinard replied:
Got it. Thanks for the info!

etpinard commented on test/image/mocks/bar_attrs_relative.json:
```diff
@@ -2,27 +2,37 @@
         {
             "width":[1,0.8,0.6,0.4],
+            "text":[1,2,3333333333,4],
+            "textposition":"outside",
```
If I understand correctly, only the outer most bar can have outside text?

n-riesco replied:
Yes, that's correct. Is there anything wrong?

etpinard replied:
Nothing wrong. Just double checking. :+1:

etpinard commented on test/jasmine/tests/bar_test.js:
```diff
@@ -623,13 +871,13 @@ describe('A bar plot', function() {
             assertPointField(cd, 'y', [
                 [1, 2, 3, 4], [4, 4, 4, 4],
-                [-1, -3, -2, -4], [4, -4, -5, -6]]);
+                [-1, -3, -2, -4], [4, -3.25, -5, -6]]);
```
why?

n-riesco replied:
I wanted to have a small negative bar, so that the label `"outside text"` doesn't fit inside the bar and has to be drawn below the bar (as opposed to the positive bar case that has outside texts drawn above the bar).

etpinard replied:
got it, thanks!

etpinard commented on test/jasmine/tests/bar_test.js:
```diff
@@ -674,6 +922,88 @@ describe('A bar plot', function() {
         });
     });
+
+    it('should coerce text-related attributes', function(done) {
```
Nice test. Can you make sure that `Plotly.restyle` works with the new attributes too?

etpinard replied:
:arrow_up: **blocking** before making a PR to plotly.js

alexcjohnson commented on test/image/mocks/bar_attrs_relative.json:
```diff
@@ -2,27 +2,37 @@
             "type":"bar"
         }, {
             "width":[0.4,0.6,0.8,1],
+            "text":["Three",2,"outside text",0],
```
Two questions here:
1. why is "outside text" outside? Wouldn't it be just as big on the inside?
2. Where did "0" go? Even though the bar has no size, because it's the last bar we _could_ show it where the last bar would have been (and likewise, it's confusing that the "4" from trace 0 shows up on the outside, it feels like only trace 1 should be able to have a label there.

n-riesco replied:
> Why is "outside text" outside? Wouldn't it be just as big on the inside?

The way I understood the criteria for `auto` is that if it doesn't fit inside, it goes outside (no matter if shrinking is required to draw it outside).

> Where did "0" go?

0-size bars aren't even included in `gd.calcdata`

> it feels like only trace 1 should be able to have a label there

I think this feeling is specific to this case, because `trace[1]` only has bars with positive sizes. If `trace[1][3]` was `-1` instead of `0`, then the bar and the label would be drawn on the negative side.

alexcjohnson replied:
> The way I understood the criteria for auto is that if it doesn't fit inside, it goes outside (no matter if shrinking is required to draw it outside).

Yeah, that's probably what I said :) But would you agree that if the shrinking is the same either way then it should stay inside?

> 0-size bars aren't even included in gd.calcdata

OK cool - yes, definitely an edge case, lets leave this one as it is.

n-riesco replied:
@alexcjohnson OK, I'll change `auto`:
- if it fits inside, the label will be drawn inside
- if it doesn't, the label will be drawn wherever is larger

alexcjohnson commented on test/image/mocks/bar_attrs_group_norm.json:
```diff
@@ -1,19 +1,24 @@
         }, {
-            "base":[7,6,5,4],
-            "x":[1,2,3,4],
+            "text":0.5,
```
It's confusing to read this plot with the different text labels rotated different directions. I'd rotate them only clockwise (if they're going to rotate)

n-riesco replied:
How about we don't rotate the labels at all and let the user decide?

Something like this:

```
    textrotation: {
        valType: 'number',
        dflt: null,
        arrayOk: true,
        role: 'info',
        description: [
            'Rotates the bar text (in clockwise degrees).'
        ].join(' ')
    },
```

etpinard replied:
Adding `textrotation` isn't a bad idea at all. I'm sure some users will want to rotate their labels to 45 degrees.

alexcjohnson replied:
That could be an option - for conformity with tick labels it should be `textangle` - though offhand I'd say it opens more questions than it answers. For example, you might want 45 degrees or something for text on vertical bars in order to make the labels more readable, but have the labels expand into the space of the next bar(s) - similar to how angles are used on tick labels. But then because these labels start from different vertical points, making it so they don't overlap becomes challenging.  I would leave this out for now, and then when we do add it, this will override the "auto" rotation that we determine now.

So then thinking a little more about what this auto behavior should be - basically I feel like text should be horizontal unless it can be less shrunken being vertical - similar to the inside/outside criterion. I think that's already what you're doing for vertical bars, but for horizontal it looks like you're defaulting to vertical text when I think 99% of the time it should be horizontal, and then in those cases that you _do_ decide to rotate the text, it should only turn 90 degrees clockwise.

n-riesco replied:
@alexcjohnson For vertical bars, the criteria of maximising label size, would cause the `outside text` labels in the plot below to be vertical:

![image](https://cloud.githubusercontent.com/assets/6199391/20326189/9bdb64d4-ab7f-11e6-8af8-b388c383633b.png)

Would that be OK?

etpinard commented on src/traces/bar/attributes.js:
```diff
@@ -35,8 +42,32 @@ module.exports = {
+
     text: scatterAttrs.text,
 
+    textposition: {
```
@n-riesco what are your thoughts on adding a `textinfo` attribute (similar to the pie) down the road? 

The default value would be `'text'` which corresponds the value linked to the `text` attribute. Another possiblevalues could be `'total'` (i.e. the total in a `stack` barmode).

n-riesco replied:
What would be the meaning of `'total'` for inner bars? no label? the bar size? the cum sum up to that bar?

etpinard replied:
> What would be the meaning of 'total' for inner bars? 

I'm thinking `'total'` would draw only one text item per bar group. But yeah, this would open up a cross-trace problem. Not so fun.

n-riesco replied:
@etpinard I think this is doable. I can compute the total at the some point I determine which bars are the outmost bars.

etpinard replied:
>  I think this is doable

Great. That's all I wanted to here for now. No action required in this iteration.

etpinard commented on src/traces/bar/attributes.js:
```diff
@@ -35,8 +42,32 @@ module.exports = {
+        dflt: 'none',
+        arrayOk: true,
+        description: [
+            'Specifies the location of the `textinfo`.'
```
@n-riesco this would be a great place to add info about the inside / outside algorithm you implemented.

etpinard replied:
:arrow_up: **blocking** before making a PR to plotly.js

n-riesco committed:
test: update bar restyle test

* Tested that `textposition` can also be restyled.

n-riesco committed:
bar: bar texts only rotate clockwise

n-riesco committed:
test: update mock bar_attrs_group_norm

* Changed bar texts to "50%" to ensure the bounding box width is larger
  than its height (unfortunately, this wasn't case when the baseline
  images were generated; the height of the bounding box includes some
  white space).

* Swapped the `texposition` of both traces so that illustrates the least
  favourable case for clockwise rotation.

n-riesco committed:
test: update baseline images

n-riesco commented:
@alexcjohnson I've updated the rotation algorithm. See the images below:
- now rotations are always 90 degrees clockwise
- in the case of outside labels, I'm still preventing rotation if this results in a label being perpendicular to the bar
- I've changed the labels in the first plot from "0.5" to "50%" to ensure that the text width is larger than the height (unfortunately, this wasn't case in the baseline images, and it was hard to reason why the inside labels had been rotated)
- I've also swapped textposition in the first plot, so that trace1 shows rotated outside labels (the case I think is the least favourable for clockwise rotation)

Questions before I continue:
- do you like the results of the new criteria?
- do you still think that I should change the criteria for the `'auto'` mode?

![](https://raw.githubusercontent.com/n-riesco/plotly.js/8cefb5be17d5ac4b2a92f1e9759fec27f7102a97/test/image/baselines/bar_attrs_group_norm.png)

![](https://raw.githubusercontent.com/n-riesco/plotly.js/8cefb5be17d5ac4b2a92f1e9759fec27f7102a97/test/image/baselines/bar_attrs_relative.png)

alexcjohnson commented:
@n-riesco the rotation algorithm is better but I still think for both horizontal and vertical bars the first choice should be horizontal text, and you only rotate it if horizontal is squished and vertical is (less) squished. That means that really the _only_ time you would end up rotating text on horizontal bars is when inside is explicitly specified and the bar size is small enough that it's thinner than a square.

And again, in `bar_attrs_relative` I don't see why "outside text" in trace 1 is outside the orange bar, it looks like it would be just as big inside the bar, wouldn't it?

n-riesco commented:
@alexcjohnson Thx, that's helped clear my mind a lot.

n-riesco committed:
bar: simplify outside text positioning

* Don't rotate outside labels.

n-riesco committed:
bar: update criteria for 'textposition: 'auto'`

* 'auto' will draw a label inside a bar if larger than drawn outside.

n-riesco committed:
test: update bar tests

* Updated bar tests to account for the change in criteria for
  `textposition:'auto'`.

n-riesco committed:
bar: describe `textposition` options

n-riesco committed:
test: update baseline images

n-riesco commented:
These are the latest results using these criteria:
- `auto` only draws outside if the result is larger than inside
- `inside` labels are rotated if it prevents excesive shrinkage
- `outside` labels are never rotated
- rotation are always 90 degrees clockwise

![](https://raw.githubusercontent.com/n-riesco/plotly.js/cdc298da36c454dbdfed287a596986a1d2e4ba78/test/image/baselines/bar_attrs_group_norm.png)

![](https://raw.githubusercontent.com/n-riesco/plotly.js/cdc298da36c454dbdfed287a596986a1d2e4ba78/test/image/baselines/bar_attrs_relative.png)
